### ISSUE TITLE
0.31 Custom expression aggregations throw JS errors and won't render on refresh

### ISSUE THREAD

**Issue by nickhamlin** [metabase/metabase#8940]
Your browser and the version: Chrome 70.0.3538.67 (Official Build) (64-bit)
Your operating system: OS X 10.12.6
Your databases: MySQL
Metabase version: 0.31.0
Metabase hosting environment: Docker
Metabase internal database: MySQL

*Repeatable steps to reproduce the issue*
I'm noticing issues saving questions that use many of the new table formatting features simultaneously. I haven't yet been able to isolate exactly which, if any, individual features are causing the issues, but I believe it's related to the fact that the question isn't getting saved/cached properly.  In this question I've
- Added some filters
- Grouped by a column
- summed a custom field, along with some preexiting fields
- Renamed the columns to something other than "sum"
- Applied conditional formatting to one column
- Added mini bars for another column

Watching the logs when I save the question provides a clue. The first line below is for a simpler question that works fine.  Immediately after, I tried to save my more complex question. I know that these are the POSTs associated with each since I had the logs open on a second screen as I hit the button. 

Nov 14 20:41:08 DEBUG metabase.middleware :: POST /api/card/1275/query 200 (113 ms) (15 DB calls). Jetty threads: 8/50 (4 busy, 8 idle, 0 queued)
Nov 14 20:43:55 DEBUG metabase.middleware :: POST /api/card 200 (121 ms) (45 DB calls). Jetty threads: 8/50 (3 busy, 7 idle, 0 queued)

It looks like the first is POSTing a path with a unique ID (1275) in it, while the second isn't (it should've been 1276, which is what appears in my browser URL bar after I save. However, if I try to open that same 1276 URL in another window, I get a blank screen (just the blue metabase bar at the top, no query builder or results). The same thing happens if I copy/paste the unsaved URL with the question hash in it and try to open that in another window.  

Any ideas on how we can prevent this? We LOVE the new viz features and would like to use them ASAP once we can save them effectively.  Many thanks!

**Comment by tlrobinson** [comment]
@nickhamlin Do you see any errors in the JavaScript console? I tried following your steps and couldn't reproduce it.

FYI  `POST /api/card/1275/query` is running the actual query for the previously saved (1275) question, while `POST /api/card` is creating a new question. The new question doesn't have an ID until that request completes so that's why you don't see 1276. Could you see what's returned by `GET /api/card/1276`?

**Comment by nickhamlin** [comment]
@tlrobinson Thanks for investigating, I do see errors in the JS console when I try to load `/question/1276` in my browser (which I assume is just a `GET /api/card/1276` under the hood). I have three lines that look pretty similar: 

```
Uncaught (in promise) field with ID does not exist: 18896
Uncaught (in promise) field with ID does not exist: 18896
Uncaught field with ID does not exist: 18896
```
This made me suspect the bug is related to renaming of fields.  After more poking around, I've been able to isolate it to *situations where a new title is applied to a column containing a custom expression*.

**Comment by nickhamlin** [comment]
Also, to clarify the title on this after doing more digging, I think the question is getting saved properly. Instead, the issue is trying to render it when it's accessed again. This also explains why it's happening with both saved and unsaved questions.

**Comment by tlrobinson** [comment]
@nickhamlin I'm still having trouble reproducing this. Could you try reproducing it using a table in the "Sample Dataset" database?

**Comment by nickhamlin** [comment]
@tlrobinson You bet, here are the steps:

1. Choose "Orders" as data source
2. Make a custom expression. I used `Sum(Total - Tax)`
3. Group by something. I used `Product -> Category`
4. Run the question so the table appears
5. Choose "formatting" from the heading of the second column (the one with the aggregated custom expression) and give this column a new custom title.
6. Save the question, replacing the default title with something else.
7. Try to access the saved question in a new browser window to throw the JS errors.

I get the same errors I mentioned above when doing this, but with a different number at the end (which makes sense, since I'm looking at a different field).  Please let me know if you need any more info!

**Comment by tlrobinson** [comment]
@nickhamlin Thank you, I have reproduced it and will work on a fix.

**Comment by tlrobinson** [comment]
Actually it looks like this doesn't have anything to do with changing the column name. Simply adding any custom aggregation and hitting refresh causes the same issue. I suspect it's because the metadata for the field in the custom expression isn't loaded.

**Comment by camsaul** [comment]
Fixed by #8989
